lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Delegation.hs

```diff
@@ -0,0 +1,361 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Cardano.Wallet.Primitive.AddressDiscovery.Delegation
+    ( -- * Creation
+      DelegationState (..)
+    , initialDelegationState
+    , unsafeDeserializeDelegationState
+
+      -- * Operations
+    , presentableKeys
+    , usableKeys
+    , activeKeys
+
+    -- * For Testing
+    , keyAtIx
+    , lastActiveIx
+    , PointerUTxO (..)
+
+      -- * Chain following model
+    , Tx (..)
+    , Cert (..)
+    , applyTx
+    , setPortfolioOf
+    )
+    where
+
+import Prelude
+
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , DerivationType (..)
+    , Index (..)
+    , MkKeyFingerprint (paymentKeyFingerprint)
+    , Role (..)
+    , SoftDerivation (..)
+    , ToRewardAccount (..)
+    )
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address )
+import Cardano.Wallet.Primitive.Types.Coin
+    ( Coin (..) )
+import Cardano.Wallet.Primitive.Types.Hash
+    ( Hash (..) )
+import Cardano.Wallet.Primitive.Types.RewardAccount
+    ( RewardAccount )
+import Cardano.Wallet.Primitive.Types.Tx
+    ( TxIn (..), TxOut (..) )
+import Control.DeepSeq
+    ( NFData )
+import Data.Maybe
+    ( maybeToList )
+import GHC.Generics
+    ( Generic )
+import Quiet
+    ( Quiet (..) )
+
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TB
+
+--------------------------------------------------------------------------------
+-- Stake Key State
+--------------------------------------------------------------------------------
+
+data DelegationState k = DelegationState
+    { -- | The account public key from which the stake keys should be derived.
+      rewardAccountKey :: k 'AccountK XPub
+
+      -- | The next unregistered stake key.
+      --
+      -- The active stake keys can be described as the half-open interval
+      --    @@ [0, nextKeyIx ) @@
+      -- having a length of @nextKeyIx@.
+      --
+      -- E.g. a state with nextKeyIx=0 has no keys, and nextKeyIx=2 has the keys
+      -- [0,1].
+    , nextKeyIx :: !(Index 'Soft 'AddressK)
+
+      -- | By threading a long a "pointer UTxO", we leverage the ledger rules to
+      -- ensure transactions cannot be re-ordered in a way that breaks the
+      -- rules/assumtions for how stake-keys are managed.
+    , pointer :: !(Maybe PointerUTxO)
```

Comment:
This `PointerUTxO` now also contains the `nextKeyIx`, so we _should_ be able to remove one of them. I think.

Reply:
Removed in cc1f2bc